src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { S3 } from 'aws-sdk'
+import { ListObjectsV2Output } from 'aws-sdk/clients/s3'
+import * as _ from 'lodash'
+import * as vscode from 'vscode'
+import * as fs from 'fs'
+import { basename } from 'path'
+import * as stream from 'stream'
+import { ext } from '../extensionGlobals'
+import { promisifyReadStream, promisifyWriteStream } from '../streamUtilities'
+import {
+    Bucket,
+    CreateBucketRequest,
+    CreateFolderRequest,
+    DEFAULT_DELIMITER,
+    DEFAULT_MAX_KEYS,
+    Folder,
+    DownloadFileRequest,
+    ListBucketsResponse,
+    ListObjectsRequest,
+    ListObjectsResponse,
+    File,
+    S3Client,
+    S3Error,
+    UploadFileRequest,
+    CreateFolderResponse,
+    CreateBucketResponse,
+} from './s3Client'
+import * as mime from 'mime-types'
+
+const DEFAULT_CONTENT_TYPE = 'application/octet-stream'
+const BUCKET_REGION_HEADER = 'x-amz-bucket-region'
+
+export class DefaultS3Client implements S3Client {
+    public constructor(
+        private readonly regionCode: string,
+        private readonly partitionId: string,
+        private readonly s3Provider: (regionCode: string) => Promise<S3> = createSdkClient,
+        private readonly fileStreams: FileStreams = new DefaultFileStreams()
+    ) {}
+
+    public async createS3(): Promise<S3> {
+        return this.s3Provider(this.regionCode)
+    }
+
+    /**
+     * Creates a bucket in the region of the client.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    public async createBucket(request: CreateBucketRequest): Promise<CreateBucketResponse> {
+        const s3 = await this.createS3()
+
+        try {
+            // The docs incorrectly state that you must always specify the LocationConstraint or it will default to us-east-1
+            // https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html
```

Comment:
I don't see this as a reason to not be explicit about it

Reply:
That's fair. I previously submitted them doc feedback, so I'll make it explicit here.